### PR TITLE
CachingHiveMetastore: refactor cache creation and flushCache()

### PR DESCRIPTION
**Pull request by iirekm**

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 1]
```diff
@@ -160,96 +163,77 @@ protected CachingHiveMetastore(HiveMetastore delegate, Executor executor, Option
...
-                        return loadPartitionByName(partitionName);
-                    }
-
-                    @Override
-                    public Map<WithIdentity<HivePartitionName>, Optional<Partition>> loadAll(Iterable<? extends WithIdentity<HivePartitionName>> partitionNames)
-                    {
-                        return loadPartitionsByNames(partitionNames);
-                    }
-                }, executor));
+        class CacheFactory
```
Comment (iirekm) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:166]: mixture of factory method and method object patterns

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 2]
```diff
@@ -160,96 +163,77 @@ protected CachingHiveMetastore(HiveMetastore delegate, Executor executor, Option
...
-        partitionNamesCache.invalidateAll();
-        databaseCache.invalidateAll();
-        tableCache.invalidateAll();
-        partitionCache.invalidateAll();
-        partitionFilterCache.invalidateAll();
-        tablePrivilegesCache.invalidateAll();
-        tableStatisticsCache.invalidateAll();
-        partitionStatisticsCache.invalidateAll();
-        rolesCache.invalidateAll();
+        caches.forEach(Cache::invalidateAll);
```
Comment (iirekm) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:236]: it's nothing new, it's very similar to `Closer` already used in many places, the difference is that here objects can be 'closed' many times

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 3]
```diff
@@ -114,6 +116,7 @@
     private final LoadingCache<HivePrincipal, Set<RoleGrant>> roleGrantsCache;
     private final LoadingCache<String, Set<RoleGrant>> grantedPrincipalsCache;
     private final LoadingCache<String, Optional<String>> configValuesCache;
+    private final List<LoadingCache<?, ?>> caches = new ArrayList<>();
```
Comment (losipiuk) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:119]: Initialize in constructor to `ImmutableList`

**Comment by iirekm** [comment]
Benefit is: all caches are flushed (previously 4 were missing), code is
more readable, avoids errors that somebody forgot to call caches.add

wt., 18 sie 2020, 15:30 użytkownik Łukasz Osipiuk <notifications@github.com>
napisał:

**Comment by losipiuk** [comment]
> Benefit is: all caches are flushed (previously 4 were missing), code is more readable, avoids errors that somebody forgot to call caches.add

Fine :) THanks

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 2 above)
Comment (electrum) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:239]: Nit: We prefer normal for-each loops. Use `foreach` if there's a specific benefit to doing so, such as the terminal operation of a stream. Note that `Map.forEach` is far superior to for-each as it avoids unpacking the `Entry`.

**Comment by electrum** [comment]
For the commit message, we write the title like a phrase or sentence, rather than prefixing with a component. So something like

> Refactor Hive cache creation and flushing

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 4]
```diff
@@ -160,96 +163,80 @@ protected CachingHiveMetastore(HiveMetastore delegate, Executor executor, Option
...
-                        return loadPartitionByName(partitionName);
-                    }
-
-                    @Override
-                    public Map<WithIdentity<HivePartitionName>, Optional<Partition>> loadAll(Iterable<? extends WithIdentity<HivePartitionName>> partitionNames)
-                    {
-                        return loadPartitionsByNames(partitionNames);
-                    }
-                }, executor));
+        class CacheFactory
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:166]: please extract private static inner class. Class definition inside the method body is hard to follow.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java [hunk 5]
```diff
@@ -13,6 +13,8 @@
  */
```
Comment (kokosing) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:13]: WRT to commit message. It looks like the refactor is not the most important change here. The most important part is that you added missing invalidate for caches. And refactor is only about preventing this to happen in future again. Please consider writing better commit message.

**Review by kokosing — COMMENTED** [review]
Looks good % comments

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 4 above)
Comment (iirekm) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:166]: If it was deeper, it would be a good candidate for moving "up"; everything not too nested (let's say up to 4-5 nested `{` / `}`) is not a readability problem on today's wide screens; it even improves readability (moving this "up" would mean some boilerplate like explicit field and constructor declarations).

I think it's possible to configure Checkstyle such way: allow up to some 5 nested `{`s, block more than that.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 5 above)
Comment (iirekm) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:13]: couldn't find a better message: it's not just a bugfix, but also refactoring to remove duplication

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 4 above)
Comment (iirekm) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:166]: in intellij it's possible to configure in `Inspections > Class metrics / Method metrics`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java (hunk 5 above)
Comment (kokosing) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/cache/CachingHiveMetastore.java:13]: > it's not just a bugfix, but also refactoring to remove duplication

So maybe two commits. One bugfix. Second refactofing. WDYT?

**Review by kokosing — APPROVED** [review]

**Comment by colebow** [comment]
:wave: @iirekm - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
